src/main/scala/org/clulab/wm/eidos/groundings/grounders/EidosOntologyGrounder.scala

```diff
@@ -55,37 +66,91 @@ abstract class EidosOntologyGrounder(val name: String, val domainOntology: Domai
-    groundPatternsThenEmbeddings(text, text.split(" +"), patterns, embeddings)
+  // For Regex Matching
+  def nodesExampleMatched(string: String, nodes: Seq[ConceptExamples]): Map[Namer, Float] = {
+    (
+      for (node <- nodes)
+      yield node.namer -> nodeExamplesMatch(string, node.examples)
+    ).toMap
   }
-  def groundPatternsThenEmbeddings(splitText: Array[String], patterns: Seq[ConceptPatterns], embeddings: Seq[ConceptEmbedding]): MultipleOntologyGrounding = {
-    groundPatternsThenEmbeddings(splitText.mkString(" "), splitText, patterns, embeddings)
+
+  def nodeExamplesMatch(string: String, examples: Option[Array[String]]): Float = {
+    examples match {
+      case None => string.length.toFloat
+      case Some(examples) =>
+        val lowerString = string.toLowerCase // just once for all examples
+        examples
+            .map { example => new EditDistance().score(lowerString, example.toLowerCase) }
+            .min
+            .toFloat
+    }
+  }
+
+  def groundExamplesThenEmbeddings(text: String, examples: Seq[ConceptExamples], embeddings: Seq[ConceptEmbedding]): MultipleOntologyGrounding = {
+    ???
   }
-  def groundPatternsThenEmbeddings(text: String, splitText: Array[String], patterns: Seq[ConceptPatterns], embeddings: Seq[ConceptEmbedding]): MultipleOntologyGrounding = {
+
+  def groundPatternsThenEmbeddings(text: String, patterns: Seq[ConceptPatterns], examples: Seq[ConceptExamples], embeddings: Seq[ConceptEmbedding]): MultipleOntologyGrounding = {
+    groundPatternsThenEmbeddings(text, text.split(" +"), patterns, examples, embeddings)
+  }
+
+  def groundPatternsThenEmbeddings(splitText: Array[String], patterns: Seq[ConceptPatterns], examples: Seq[ConceptExamples], embeddings: Seq[ConceptEmbedding]): MultipleOntologyGrounding = {
+    groundPatternsThenEmbeddings(splitText.mkString(" "), splitText, patterns, examples, embeddings)
+  }
+
+  def exactMatchForPreds(splitText: Array[String], embeddings: Seq[ConceptEmbedding]): Option[MultipleOntologyGrounding] = {
+    // This looks for exact string overlap only!
+    val joinedText = splitText.mkString(" ")
+    val lowerText = joinedText.toLowerCase
+    // text contains node name
+    val overlapWithText = embeddings.filter(embedding => lowerText.contains(StringUtils.afterLast(embedding.namer.name.toLowerCase.replaceAll("_", " "), '/', true))).filter(emb => StringUtils.afterLast(emb.namer.name.toLowerCase.replaceAll("_", " "), '/', true).nonEmpty)
+    // node name contains text
+    val overlapWithNodeName = embeddings.filter(embedding => StringUtils.afterLast(embedding.namer.name.toLowerCase.replaceAll("_", " "), '/', true).contains(lowerText))
+    // get the maximal overlap and return it
+    val overlaps = overlapWithText++overlapWithNodeName
+    val maxOverlap = if (overlaps.nonEmpty) overlaps.maxBy(_.namer.name.length) else null
```

Comment:
You are correct. We want to sort by max overlap of the trailing_name, not the whole path.

Reply:
Just now it looks like this:
```scala
  def exactMatchForPreds(splitText: Array[String], embeddings: Seq[ConceptEmbedding]): MultipleOntologyGrounding = {
    // This looks for exact string overlap only!
    val lowerText = splitText.mkString(" ").toLowerCase
    // This tuple is designed so that Seq.min gets the intended result, the one with the min negLength
    // (or max length) and in case of ties, the min canonicalName, so the first in alphabetical order.
    val overlapTuples = embeddings.flatMap { embedding =>
      val canonicalName = embedding.namer.canonicalName
      // Non-leaf nodes end with a / resulting in an empty canonicalName which we don't want to match.
      // Text contains node name.
      if (canonicalName.nonEmpty && lowerText.contains(canonicalName))
        Some(-canonicalName.length, canonicalName, embedding.namer)
      // Node name contains the text
      else if (canonicalName.contains(lowerText))
        Some(-lowerText.length, canonicalName, embedding.namer)
      else None
    }
    val multiOntologyGrounding = Collection.minOption(overlapTuples)
        .map { overlapTuple => Seq(SingleOntologyNodeGrounding(overlapTuple._3, 1.0f)) }
        .getOrElse(Seq.empty)

    multiOntologyGrounding
  }
```

Reply:
That looks good to me!

Reply:
In this method, I'm checking if one string is contained within another, as did the original code.  That containment does not have to heed word boundaries on the longer string.  There could be partial words remaining on either side of the match.  Say there was text "unconcerned" and the embedding was "concern".  There would be a match from contains, but when I go to remove matches to get the remaining content, either "concern" will not be removed because it matches no word or there might be two leftover words "un" and "ed",  United Nations and education.  It won't happen often enough to spoil the results, but it's messy and incorrect.  I'm thinking that the tokenized texts, Array[String], should be instead compared via indexOfSlice so that only whole words match and only whole words are removed.  If we want to account for partial matches that don't include prefixes or suffixes, then that can be handled explicitly when the whole words are compared through something like stemming.

Reply:
I agree that it would be better to do compare whole words only to avoid those potential issues where we might accidentally end up with leftover word parts.